Mathlib/GameTheory/Auction/Secondprice.lean

```diff
@@ -0,0 +1,290 @@
+
+## Main Results
+
+- `utility_nneg`: utility is non-negative if the bid equals the valuation.
+- `valuation_is_dominant`: Bidding one's valuation is a dominant strategy.
+- `first_price_has_no_dominant_strategy`: There is no dominant strategy in a first price auction.
+
+## Helping Lemmas
+
+- `gt_wins`: If `i`'s bid is higher than all other bids, then `i` wins.
+- `exists_max`: There exists a participant whose bid matches the highest bid
+- `winner_take_max`: The winner's bid is the highest.
+- `b_winner`: The winner's bid is at least the second highest bid.
+- `utility_winner`: If `i` wins, utility is the valuation minus the second highest bid.
+- `utility_loser`: If `i` does not win, their utility is 0.
+- `b_winner_max`: The winner's bid is greater than or equal to all other bids.
+- `b_loser_max`: If `i` does not win, the highest bid excluding `i` matches the highest bid.
+- `utility_first_price_winner`: If `i` wins in a first price auction utility is valuation minus bid.
+- `utility_first_price_loser`: If `i` does not win in a first price auction, utility is 0.
+
+## Notations
+
+- `|b|`: Represents a bidding function.
+- `maxb(b)`: The highest bid in the function `b`.
+- `B i`: The maximal bid of all participants but `i`.
+
+## Implementation Notes
+
+The structure and functions assume the existence of multiple bidders to allow for meaningful
+auction dynamics. Definitions like `winner` and `maxb` make use of Lean's `Finset` and `Classical`
+logic to handle potential non-constructive cases effectively.
+## References
+
+* [T. Roughgarden, *Twenty lectures on Algorithmic Game Theory*][roughgarden2016]
+
+## Tags
+
+auction, game theory, economics, bidding, valuation
+-/
+
+open Classical
+
+/-!### Structure Definition -/
+
+/-- The Auction structure is set with components including: -/
+structure Auction where
+   /-- A set of participants.-/
+   I : Type*
+   /-- A `Fintype` instance. -/
+   hF : Fintype I
+   /-- An `Inhabited` instance. -/
+   hI : Nontrivial I
+   /-- A function mapping each participant to their valuation. -/
+   v : I → ℝ
+
+attribute [instance] Auction.hF Auction.hI
+
+namespace Auction
+
+variable {a : Auction} (b : a.I → ℝ)
+
+/-!### Helper Functions and Definitions -/
+/-- Computes the highest bid given a bidding function `b`. -/
+@[simp]
+def maxb : ℝ := Finset.sup' Finset.univ Finset.univ_nonempty b
+
+/-- A strategy is dominant if bidding `bi` ensures that
+`i`'s utility is maximized relative to any other bids `b'` where `b i = bi`. -/
+def dominant (utility : (a.I → ℝ) → a.I → ℝ) (i : a.I) (bi : ℝ) : Prop := ∀ b b' : a.I → ℝ,
+  (b i = bi) → (∀ j : a.I, j ≠ i → b j = b' j)
+   → utility b' i ≤ utility b i
+
+/-- `hole` represents a bidding function by setting the bid of participant `i` to `bi`, while other
+bids unchanged. -/
+noncomputable abbrev hole (i : a.I) (bi : ℝ) (b : a.I → ℝ) : a.I → ℝ :=
+  fun j => if j=i then bi else b j
```

Comment:
This already exists in the mathlib. It's called `Function.update`.
```suggestion
  Function.update b i bi
```

Reply:
It is a huge shock to me. Thanks for your suggestion!